### PR TITLE
APEXCORE-700 Uniform interface between setup and runtime plugins

### PR DESCRIPTION
**Pull request by pramodin**
@tushargosavi  please see
Better to merge after https://github.com/apache/apex-core/pull/503/

### CODE REVIEW

**Comment by pramodin** [comment]
rebased

File: engine/src/main/java/com/datatorrent/stram/plan/logical/DAGSetupPluginManager.java [hunk 1]
```diff
@@ -37,7 +44,10 @@
   private Configuration conf;
 
   public static final String DAGSETUP_PLUGINS_CONF_KEY = "apex.plugin.dag.setup";
-  private DAGSetupPlugin.DAGSetupPluginContext contex;
+
+  private PluginInfoMap<DAGSetupPlugin> pluginInfoMap = new PluginInfoMap<>();
+
+  //private Map<EventType, List<Plugin.EventHandler>> eventHandlers = Maps.newHashMap();
```
Comment (tushargosavi) [inline engine/src/main/java/com/datatorrent/stram/plan/logical/DAGSetupPluginManager.java]: Remove commented code.

**Review by tushargosavi — APPROVED** [review]
LGTM ..

File: engine/src/main/java/com/datatorrent/stram/plan/logical/DAGSetupPluginManager.java (hunk 1 above)
Comment (pramodin) [reply engine/src/main/java/com/datatorrent/stram/plan/logical/DAGSetupPluginManager.java]: Done

**Comment by pramodin** [comment]
Anyone else who wants to review it, please do so, so that it can be ready for the release.

**Comment by vrozov** [comment]
@PramodSSImmaneni let me take a look

File: api/src/main/java/org/apache/apex/api/plugin/Plugin.java [hunk 2]
```diff
@@ -21,12 +21,36 @@
...
+  {
+    /**
+     * Handle a event.
+     *
+     * This method is called when the event occurs.
+     *
+     * @param type
+     * @param data
+     */
+    void handle(EventType<T> type, T data);
```
Comment (vrozov) [inline api/src/main/java/org/apache/apex/api/plugin/Plugin.java]: Encapsulate `EventType<T> type` and `T data` into a single object `Event event` that has Type and optionally `data` specific for that Event.

Comment (pramodin) [reply api/src/main/java/org/apache/apex/api/plugin/Plugin.java]: Put them into a single data structure.

**Comment by pramodin** [comment]
Incorporated suggestions from Vlad's PR https://github.com/PramodSSImmaneni/apex-core/pull/2 and updated

**Review by vrozov — APPROVED** [review]
LGTM

File: engine/src/main/java/org/apache/apex/engine/api/plugin/DAGExecutionPlugin.java [hunk 3]
```diff
@@ -18,26 +18,66 @@
...
+ *   <li>{@see Context.COMMIT_EVENT} When committedWindowId changes in the platform an event will be delivered to the plugin</li>
  * </ul>
  *
- * A plugin should register a single handler for an event, In case multiple handlers are registered for an event,
- * then the last registered handler will be used. Plugin should cleanup additional resources created by it during shutdown
- * such as helper threads and open files.
  */
-@InterfaceStability.Evolving
-public interface DAGExecutionPlugin extends Plugin<DAGExecutionPluginContext>
+public interface DAGExecutionPlugin<T extends DAGExecutionPlugin.Context> extends Plugin<T>
```
Comment (tushargosavi) [inline engine/src/main/java/org/apache/apex/engine/api/plugin/DAGExecutionPlugin.java:47]: Needs to be marked as Evolving.

File: engine/src/main/java/org/apache/apex/engine/api/plugin/DAGExecutionPlugin.java [hunk 4]
```diff
@@ -18,26 +18,66 @@
...
+   *
+   * Following events are supported
+   * <ul>
+   *   <li>{@see Context.HEARTBEAT} The heartbeat from a container is delivered to the plugin after it has been handled by stram</li>
+   *   <li>{@see Context.STRAM_EVENT} All the Stram event generated in Stram will be delivered to the plugin</li>
+   *   <li>{@see Context.COMMIT_EVENT} When committedWindowId changes in the platform an event will be delivered to the plugin</li>
+   * </ul>
+   *
+   */
+  interface Context<E extends DAGExecutionEvent> extends PluginContext<DAGExecutionEvent.Type, E>
```
Comment (tushargosavi) [inline engine/src/main/java/org/apache/apex/engine/api/plugin/DAGExecutionPlugin.java:61]: Needs to be marked as Evolving

**Review by tushargosavi — APPROVED** [review]
LGTM, except two interfaces needs to be marked as Evolving.

File: engine/src/main/java/org/apache/apex/engine/api/plugin/DAGExecutionPlugin.java (hunk 3 above)
Comment (vrozov) [reply engine/src/main/java/org/apache/apex/engine/api/plugin/DAGExecutionPlugin.java:47]: All engine interfaces and classes are implicitly Evolving.

**Pull request merged**